### DISCUSSION TITLE
OffsetCenterOfMassShape and moment of inertia

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#397] [Q&A] [answered]
I've come across what seems (to me) to be a discrepancy between how Jolt behaves and how Godot Physics behaves, and I'm curious to hear your take on it.

I'm using Jolt's `OffsetCenterOfMassShape` to implement Godot's ability to set a custom center of mass for a body. However, I've noticed when applying torque and torque impulses to such a body that Jolt behaves as if the center of mass hasn't been offset at all, whereas Godot Physics does. Backtracking through the code I can indeed see that, given identical circumstances, the inverse inertia that gets used when applying a torque impulse is different between the two.

I set up a test scene with a body with no gravity, where its center of mass is constantly in the center of the scene, but its one and only shape (a box) oscillates back and forth along its local X-axis. On top of that I am applying a constant torque every tick.

Here is what Godot Physics looks like:

https://user-images.githubusercontent.com/4884246/211683972-a61741ec-3dd1-4cb6-b1d4-458bf622453a.mp4

Here is what Jolt looks like:

https://user-images.githubusercontent.com/4884246/211683997-be621a23-2e0e-414f-a753-97dddf76fba7.mp4

(Note that anything but the box is just a visual aid and has no bearing on the simulation.)

As you can hopefully see, Jolt ends up with a constant acceleration whereas Godot Physics does not.

So my questions are:

1. Which of the two would you consider to be the correct behavior? My intuition is really off here, and I keep ping-ponging between the two.
2. If Godot's behavior is the correct one, could I be doing something wrong?
3. If there is no correct answer, can I emulate Godot's behavior somehow? Maybe using a custom decorator shape that manipulates the mass properties?

EDIT 1: Unity/PhysX seems to behave in the same way as Godot.
EDIT 2: Unreal/Chaos seems to behave in the same way as Jolt.

**Comment by jrouwe** [comment, marked as answer]
Hello,

I think I made a mistake in OffsetCenterOfMassShape, can you try this:

```
	// See Shape::GetMassProperties
	virtual MassProperties			GetMassProperties() const override
	{
		MassProperties mp = mInnerShape->GetMassProperties();
		mp.Translate(mOffset);
		return mp;
	}
```

B.t.w. how are you changing the center of mass for this body? Are you creating a new OffsetCenterOfMassShape and replacing the previous one through BodyInterface::SetShape(body_id, new_shape, true, ...)? That's the only way the center of mass can be updated on the fly.

**Reply by mihe** [reply]
> Are you creating a new OffsetCenterOfMassShape and replacing the previous one through BodyInterface::SetShape(body_id, new_shape, true, ...)?

Yes, although I pass `false` and update the mass properties myself (in a manner similar to your `GetMassProperties` in `BodyCreationSettings`), for reasons that I can't exactly recall right now.

But yes, I'm currently rebuilding the entire set of shapes whenever you do stuff like add/remove/move shapes or change the body's center of mass. It's a crude solution, but it works for the simpler stuff I'm dealing with in my tests. I imagine it won't be so great when you have more complex sets of shapes. The Bullet implemention in Godot 3 did something similar to this, but I think they took care not to rebuild stuff unless it was necessary, or something like that.

Anyway, I can happily report that your proposed fix does indeed seem to make Jolt behave just like Godot and Unity/PhysX does! Here's what it looks like with the test scene:

https://user-images.githubusercontent.com/4884246/211919602-050fa768-ec97-4c6c-99e0-76458b16fcac.mp4

Would this fix be something you could commit to Jolt, or would it be too much of a breaking change?

**Reply by jrouwe** [reply]
Cool! I will test it a bit further and submit it.

**Reply by mihe** [reply]
So it seems the values I chose were quite fortunate, depending on how you look at it, as it now seems quite susceptible to becoming unstable in its orbit due to what I presume is compounding floating-point imprecision. If you apply enough torque, or apply it for long enough, it will start to wobble and eventually lose its orbit.

Interestingly this problem does not seem to present itself if you apply the torque using impulses.

I reproduced it using a modified `OffsetCenterOfMassShapeTest` just to make sure I wasn't doing something wrong on my end:

```cpp
void OffsetCenterOfMassShapeTest::Initialize() 
{
    // Floor
    Body &floor = CreateFloor();
    floor.SetFriction(1.0f);

    Ref<ShapeSettings> sphere = new SphereShapeSettings(1.0f);
    Ref<OffsetCenterOfMassShapeSettings> offset_shape = new OffsetCenterOfMassShapeSettings(Vec3(-5, 0, 0), sphere);

    Body &body = *mBodyInterface->CreateBody(BodyCreationSettings(offset_shape, RVec3(0, 5, 0), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING));
    body.GetMotionProperties()->SetGravityFactor(0.0f);
    body.GetMotionProperties()->SetMaxAngularVelocity(INFINITY);
    body.GetMotionProperties()->SetMaxLinearVelocity(INFINITY);
    mBody = body.GetID();
    mBodyInterface->AddBody(mBody, EActivation::Activate);
}

void OffsetCenterOfMassShapeTest::PrePhysicsUpdate(const PreUpdateParams &inParams)
{
    const float torque = 100000.0f;
    mBodyInterface->AddTorque(mBody, Vec3(0, torque, 0)); // <--- Quickly becomes unstable
    // mBodyInterface->AddAngularImpulse(mBody, Vec3(0, torque * inParams.mDeltaTime, 0)); // <--- Seems stable
}
```

I'm not sure it's something worth optimizing for necessarily, but Godot Physics doesn't seem to suffer from this, regardless of method used to apply the torque, so I figured it was worth mentioning.

**Reply by jrouwe** [reply]
I've submitted the OffsetCenterOfMassShape fix in #402.

I have also investigated why there's a difference between AddTorque and AddAngularImpulse.

AddTorque eventually calls MotionProperties::MultiplyWorldSpaceInverseInertiaByVector(...) to convert the torque to a delta velocity, while AddAngularImpulse uses MotionProperties::GetInverseInertiaForRotation(...) * vector. The former is an optimized function that I just forgot to use in AddAngularImpulse.

If you change this code in MultiplyWorldSpaceInverseInertiaByVector:

```Mat44 rotation = Mat44::sRotation(inBodyRotation * mInertiaRotation);```

to

```Mat44 rotation = Mat44::sRotation(inBodyRotation) * Mat44::sRotation(mInertiaRotation);```

(so converting 2 quaternions to matrix before multiplying instead of multiplying 2 quaternions and then converting to matrix)

Then your example is stable. Unfortunately that makes the function about 20% slower and as it is called a lot, it makes the simulation slower. Personally, I don't think it is likely that anyone will make a scene like this so I'm actually tempted to break the other case as well and use MultiplyWorldSpaceInverseInertiaByVector in AddAngularImpulse.

**Reply by mihe** [reply]
> I've submitted the OffsetCenterOfMassShape fix in #402.

Awesome, thank you for the quick resolution!

> Personally, I don't think it is likely that anyone will make a scene like this

Agreed.

> I'm actually tempted to break the other case as well and use MultiplyWorldSpaceInverseInertiaByVector in AddAngularImpulse.

I say go for it. I'd rather see them be consistent in behavior than be optimized for the contrived use-case above.

**Reply by jrouwe** [reply]
> I say go for it. I'd rather see them be consistent in behavior than be optimized for the contrived use-case above.

Done in #404

**Reply by akb825** [reply]
From what I understand, translating the inertia effectively changes the frame of reference for the inertia, but conceptually keeps it the same object properties. For example, when looking at the [list of moments of inertia](https://en.wikipedia.org/wiki/List_of_moments_of_inertia), the entry for a solid cone has separate factors based on whether the axis passes through the tip, base, or center of mass. Calling Translate() on the mass properties can convert between those different relative axes, since the overall inertia of the cone is the same, you're just changing the frame of reference.

If you wanted to adjust the center of mass for situations like emulating a shape with non-uniform density, wouldn't you want to leave the inertia tensor the same? OffsetCenterOfMassShape also mentions the example of making a boat more stable by making it more bottom-heavy, wouldn't translating the inertia effectively undo that effect? (you move the center of mass down to make it more bottom-heavy, but translating the inertia makes it more top-heavy relative to the new reference point)

In this specific example it does seem to be more appropriate to translate the inertia based on the center of mass shape, but should this be configurable to take into account the other situations?

**Reply by jrouwe** [reply]
> wouldn't translating the inertia effectively undo that effect? (you move the center of mass down to make it more bottom-heavy, but translating the inertia makes it more top-heavy relative to the new reference point)

I don't think this is true: Shifting the center of mass down will increase the inertia tensor around the forward and side axis, so the boat will be more resistant to rotation instead of being more top heavy.

In any case, if you don't want to translate the inertia, you're free to provide an alternative inertia tensor when creating the body:

```
BodyCreationSettings.mOverrideMassProperties = EOverrideMassProperties::MassAndInertiaProvided
BodyCreationSettings.mMassPropertiesOverride = ...
```

Note that there's also BodyCreationSettings.mInertiaMultiplier which can be used to change the inertia tensor with a linear factor (which was made to be able to somewhat compensate for 'hollowness').
